moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java

```diff
@@ -98,6 +98,27 @@
     pushScope(JsonScope.EMPTY_DOCUMENT);
   }
 
+  /** Copy-constructor makes a deep copy for peeking. */
+  JsonUtf8Reader(JsonUtf8Reader copyFrom) {
+    super(copyFrom);
+
+    BufferedSource sourcePeek = copyFrom.source.peek();
+    this.source = sourcePeek;
+    this.buffer = sourcePeek.getBuffer();
+    this.peeked = copyFrom.peeked;
+    this.peekedLong = copyFrom.peekedLong;
+    this.peekedNumberLength = copyFrom.peekedNumberLength;
+    this.peekedString = copyFrom.peekedString;
+
+    // Make sure our buffer has as many bytes as the source's buffer. This is necessary because
+    // JsonUtf8Reader assumes any data it has peeked (like the peekedNumberLength) are buffered.
+    try {
+      this.source.require(copyFrom.buffer.size());
```

Comment:
`this.source` -> `sourcePeek` since it's already a local

Reply:
Fixed.